### DISCUSSION TITLE
How can I get health of rr server - grpc plugin?

### DISCUSSION THREAD

**Discussion by lanphan** [roadrunner-server/roadrunner#967] [Q&A] [answered]
Hello,

I'm using RR to provide grpc services, and it works well, thanks so much.
However, lately I get 1 issue: load balancer in k8s sends requests to rr instance which doesn't have any free worker, it leads to error "No free worker" response. I think this can be prevent if we have healthcheck for grpc service like http service I see somewhere on Internet. Would you please tell me how I set it up?

Current my setting:
* rr version: rr version 2.6.4 (build time: 2021-12-07T13:25:47+0000, go1.17.4)
* Linux ubuntu container
* my rr.yml

> rpc:
>   listen: tcp://127.0.0.1:6001
> 
> status:
>   address: 127.0.0.1:2114
> 
> server:
>   command: "/usr/local/bin/php /source_code/worker.php"
>   relay: "pipes"
>   relay_timeout: "20s"
> 
> grpc:
>   listen: "tcp://0.0.0.0:9001"
>   proto: ${PROTO}
> 
>   max_send_msg_size: 50
>   max_recv_msg_size: 50
>   max_connection_idle: 0s
>   max_connection_age: 0s
>   max_connection_age_grace: 0s
>   max_concurrent_streams: 10
>   ping_time: 1s
>   timeout: 200s
> 
>   pool:
>     num_workers: ${NUM_WORKERS}
>     max_jobs: ${MAX_JOBS}
>     allocate_timeout: ${ALLOCATE_TIMEOUT}
>     destroy_timeout: 60
> 
> logs:
>   mode: development
>   encoding: console
>   level: debug
>   output: stderr

* lsof -i -P -n | grep LISTEN

> rr      127 root    3u  IPv6  25216      0t0  TCP *:6061 (LISTEN)
> rr      127 root    9u  IPv4  27719      0t0  TCP 127.0.0.1:2114 (LISTEN)
> rr      127 root   12u  IPv4  25224      0t0  TCP *:9001 (LISTEN)
> rr      127 root   16u  IPv4  25225      0t0  TCP 127.0.0.1:6001 (LISTEN)

* rr -c .rr.yaml workers

> Workers of [grpc]:
> +---------+-----------+---------+---------+---------+--------------------+
> |   PID   |  STATUS   |  EXECS  | MEMORY  |  CPU%   |      CREATED       |
> +---------+-----------+---------+---------+---------+--------------------+
> |     134 | ready     |       0 | 55 MB   |    0.04 | 2 hours ago        |
> |     138 | ready     |       0 | 55 MB   |    0.07 | 2 hours ago        |
> +---------+-----------+---------+---------+---------+--------------------+

* curl http://127.0.0.1:2114/health?plugin=http --> Service: http not found
* curl http://127.0.0.1:2114/health?plugin=grpc --> Service: grpc not found
* curl http://127.0.0.1:2114/health?plugin=rpc --> Service: rpc not found

**Comment by rustatian** [comment, marked as answer]
Hey @lanphan, thanks for the question.
Unfortunately, the `grpc` plugin doesn't have an implementation of the statuses (provided by the `Status` plugin).
It's very easy to implement them (for the workers) for the `grpc` plugin [FR](https://github.com/roadrunner-server/roadrunner/issues/928). Feel free to thumb up for this ticket 😃 
For the `grpc` we have 2 options:
1. Native health checks: https://github.com/grpc/grpc/blob/master/doc/health-checking.md
2. Implement a `status` plugin interface and provide worker statistics with it.

Regarding your `curl` requests. You don't have an `http` plugin in the configuration, thus `health` for it returns with `Service: HTTP not found`.
RPC doesn't have statuses as well as the `grpc` as I mentioned above 😃 

Also, feel free to join our discord channel https://discord.gg/TFeEmCs

**Comment by lanphan** [comment]
Thank @rustatian.
I have another question: in case I enable http plugin, and AFAIK, pool of workers is shared for all plugins (http, grpc ...)
--> therefore, I can use /health?plugin=http to have a status of free workers for grpc?

Noted that I need to have health endpoint to check the availability of workers for now, no need to have complex check like your document (grpc service available + other checks ...)

**Reply by rustatian** [reply]
Actually, the `server` plugin, which is allocating workers is shared ([FR](https://github.com/roadrunner-server/roadrunner/issues/916)). But `http` pool is dedicated to the `http` plugin and not shared with others as well as grpc, tcp, jobs plugins pools.
For the `grpc` will be available (in the next minor release) 2 checks - native (which is described in the doc) and RR specific with the `/health?plugin=grpc`. Also, keep in mind, that for the `k8s` you need readiness checks, not healthchecks -> `/ready?plugin=http`. The difference between them is that the readiness check returns 200OK if there is at least 1 Ready to accept request worker.

**Reply by lanphan** [reply]
Thank @rustatian . I'm looking forward to your next minor version, so that I can use /ready?plugin=grpc directly

**Comment by lanphan** [comment]
My last question, @rustatian , when will you release your next minor version? :)

**Reply by rustatian** [reply]
> My last question, @rustatian , when will you release your next minor version? :)

Mid-Feb approximately 😃

**Comment by lanphan** [comment]
Super cool, thanks so much, @rustatian

**Reply by rustatian** [reply]
You are very welcome 👍🏻

**Comment by lanphan** [comment]
Hi @rustatian ,
I see that RR have new version 2.8, and it has Check grpc method.
My question is, which method I need to use to check readiness for RR service?
* Check grpc method
* or, /ready?plugin=grpc as we discussed above?

Thanks so much for your support.

**Reply by lanphan** [reply]
For your information:
* CHECK grpc method --> works, I use grpc_health_probe utility
```
/usr/bin/grpc_health_probe -addr=localhost:9001
status: SERVING
```

* ready?plugin=grpc returns "grpc not found"

**Reply by rustatian** [reply]
> For your information:
> 
> * CHECK grpc method --> works, I use grpc_health_probe utility
> 
> ```
> /usr/bin/grpc_health_probe -addr=localhost:9001
> status: SERVING
> ```
> 
> * ready?plugin=grpc returns "grpc not found"

Thanks, HTTP checks for the GRPC plugin will be in the next releases.

**Reply by rustatian** [reply]
@lanphan If that is suitable for you, you may use the grpc `watch` and `check` healthchecks at the moment.

**Reply by lanphan** [reply]
hi @rustatian ,

Check is ok now, thanks so much.

**Comment by TemKaa1337** [comment]
@rustatian 
Hi!
Is gRPC readiness check implemented as HTTP check now? like `/ready?plugin=grpc`

**Reply by rustatian** [reply]
Hi @TemKaa1337 👋 
`gRPC` has the special protocol to check the health. But, if you want to have an HTTP like health/readiness checks - feel free to create a feature request for that 😃

**Reply by TemKaa1337** [reply]
Thanks for an answer 😃
I understand that, but the reason i am asking is that i saw an example in RR doc [here](https://roadrunner.dev/docs/lab-health/current/en#activation-of-the-status-plugin) in notes section this line: `http://127.0.0.1:2114/health?plugin=http,grpc`, so i thought it should work like i wrote, but seems like it doesn't :(
So can i ask, to make it work should i implement `readiness` check service on my own using gRPC [contract](https://github.com/grpc/grpc/blob/master/doc/health-checking.md) (it's not readiness contract but still...) or use external tools like `grpc_health_probe` as was wrote above, or is it natively provided by RR and i should just make correct calls?)
Thanks a lot in advance for your answer, i am asking because i don't really understand that (:

**Reply by rustatian** [reply]
You're welcome 😃 

You have the following options:
1. You may use the regular PHP (or any language you want) grpc protoc compiler (not that provided by RR), compile this [proto](https://github.com/grpc/grpc-proto/blob/master/grpc/health/v1/health.proto) and just use it in your code on the RR's `grpc` endpoint. Most LoadBalancers (or k8s) support this way of checking the health of the service automatically.
2. Second option - create a feature request ticket and I'll implement the HTTP endpoints to check the `gRPC` plugin workers.
3. And yeah, thanks for noting that, [this is](https://roadrunner.dev/docs/lab-health/current/en#activation-of-the-status-plugin) the mistake in docs 👍 I'll fix that.

Generally, I would suggest you to go with the option 1, since this is the most convenient way of checking the health of the `gRPC` service. But would also be happy to implement the HTTP checks if you create a FR 😃

**Reply by TemKaa1337** [reply]
Thanks for reply!
Am i right by saying that currently i cannot check `grpc` workers `readiness` using any of built-in RR functions? If i understood you correctly, you are talking about `healthiness` of grpc workers, not `readiness`, but i need exactly `readiness` check (which will fail if there are all workers budy - the same as readiness in http plugin). Currently i am testing `grpc_health_probe` tool just to figure out how it works, and seems like it uses [proto](https://github.com/grpc/grpc-proto/blob/master/grpc/health/v1/health.proto) provided by you, which is basically health, not readiness check.
I think i will create a feature request for that, but i need a solution right now, so it would be great if you give me direction where to move :)

**Reply by rustatian** [reply]
> Am i right by saying that currently i cannot check grpc workers readiness using any of built-in RR functions? 

Yeah, there are no `readiness` checks atm.

If you need this right now, you may fork `grpc` plugin, copy-pase this [file](https://github.com/roadrunner-server/http/blob/master/status.go) to the `grpc` plugin root. I guess you would need only minor changes, like, some variables to rename.

Then, use [velox](https://github.com/roadrunner-server/velox) to build your own RR. Do not forget to use your `grpc` plugin fork in the `velox.toml` configuration.
Here you can find tutorials: https://roadrunner.dev/docs/customization-build/current/en

**Reply by rustatian** [reply]
You may use Docker as well: https://roadrunner.dev/docs/customization-build/current/en#building
If you have questions, you may join our Discord server: https://discord.gg/spiralphp and ping me in the RoadRunner group.

**Reply by TemKaa1337** [reply]
Thanks a lot for your answers and possible solution, you helped me a lot!)
Will try to implement it 😃

**Reply by rustatian** [reply]
@TemKaa1337 https://github.com/roadrunner-server/grpc/pull/130
Will be in `2023.3.7` release this Thursday.

**Reply by TemKaa1337** [reply]
Wow, thanks a lot!
Will try it out and return with feedback :)

**Reply by TemKaa1337** [reply]
@rustatian 
Hi!
Seems like everything is working as expected and as we want, HUGE thanks for this feature!

**Reply by rustatian** [reply]
Enjoy RR ❤️
